client/beefy/src/worker.rs

```diff
@@ -216,77 +241,225 @@ where
+
+				metric_set!(self, beefy_round_concluded, block_num);
+
+				info!(target: "beefy", "🥩 Round #{} concluded, committed: {:?}.", round.1, signed_commitment);
+
+				if self
+					.backend
+					.append_justification(
+						BlockId::Number(block_num),
+						(
+							BEEFY_ENGINE_ID,
+							VersionedFinalityProof::V1(signed_commitment.clone()).encode(),
+						),
+					)
+					.is_err()
+				{
+					// just a trace, because until the round lifecycle is improved, we will
+					// conclude certain rounds multiple times.
+					trace!(target: "beefy", "🥩 Failed to append justification: {:?}", signed_commitment);
+				}
+				self.signed_commitment_sender
+					.notify(|| Ok::<_, ()>(signed_commitment))
+					.expect("forwards closure result; the closure always returns Ok; qed.");
+
+				self.set_best_beefy_block(block_num, None);
+
+				trace!(target: "beefy", "🥩 round concluded, waking up ticker");
+				// We can ignore 'buffer full' error since it means some other task already
+				// queued up a wake message.
+				let _ = self.waker_tx.try_send(());
+			}
+		}
+	}
+
+	/// Set best BEEFY block to `block_num`.
+	///
+	/// Also sends/updates the best BEEFY block hash to the RPC worker.
+	/// If the hash is not explicitly provided in `opt_block_hash`,
+	/// this function just gets it from the `Client`.
+	fn set_best_beefy_block(
+		&mut self,
+		block_num: NumberFor<B>,
+		mut opt_block_hash: Option<<<B as Block>::Header as Header>::Hash>,
+	) {
+		if opt_block_hash.is_none() {
+			// Try to get block hash ourselves.
+			opt_block_hash = match self.client.hash(block_num) {
+				Ok(h) => h,
+				Err(e) => {
+					error!(target: "beefy", "🥩 Failed to get hash for block number {}: {}",
+						block_num, e);
+					None
+				},
+			}
+		};
+		// Update RPC worker with new best BEEFY block hash.
+		opt_block_hash.map(|hash| {
+			self.beefy_best_block_sender
+				.notify(|| Ok::<_, ()>(hash))
+				.expect("forwards closure result; the closure always returns Ok; qed.")
+		});
+		// Set new best BEEFY block number.
+		self.best_beefy_block = Some(block_num);
+		metric_set!(self, beefy_best_block, block_num);
+	}
+
+	/// Return first block number pertaining to same session as `header`.
+	fn session_start_for_header(&self, header: &B::Header) -> NumberFor<B> {
+		// Current header is also the session start.
+		if let Some(_) = find_authorities_change::<B>(header) {
+			return *header.number()
+		}
+
+		// Walk up the chain looking for the session change digest.
+		let mut parent_hash = *header.parent_hash();
+		while parent_hash != Default::default() {
+			let header = self
+				.client
+				.expect_header(BlockId::Hash(parent_hash))
+				// TODO: is this proof correct?
+				.expect("header always available when following parent hash; qed.");
+			if let Some(_) = find_authorities_change::<B>(&header) {
+				// Found the first header of the session.
+				break
+			}
+			parent_hash = *header.parent_hash();
+		}
+		*header.number()
```

Comment:
It's weird that in my local testing, starting a node late in the session doesn't sync all past headers and this loop just comes up with *last finalized number that **the node witnessed***.

Is this code incorrect? Or am I starting the late node with some incorrect config?

Reply:
Might be because of `let header` in-loop local var?